### PR TITLE
Move E2E tests to efficientgo/e2e

### PR DESCRIPTION
**Pull request by matej-g**
In order to provide a more comfortable and robust experience for out integration testing, this PR moves tests from using Bash to using the [`efficientgo/e2e`](https://github.com/efficientgo/e2e) Go framework. This applies to both metrics and logs test, as well running the setup locally (interactive test). Excludes moving load tests for now.

### CODE REVIEW

File: test/integration.sh [hunk 1]
```diff
@@ -117,8 +117,7 @@ done
       run \
       --server \
       ./test/config \
-      --ignore '*.json' \
```
Comment (ianbillett) [inline test/integration.sh:120]: Can this file not be removed now we're doing the integration tests in go?

File: test/testtls/generate.go [hunk 2]
```diff
@@ -121,12 +125,19 @@ func generateCACert(commonName string) (certBundle, error) {
 	return certBundle{cert: cert, key: key}, nil
 }
 
-func generateCert(commonName string, hosts []string, groups []string, profile string, signingConfig *config.Signing, ca []byte, caKey []byte) (certBundle, error) {
-	fmt.Printf("cert generate, commonName=%s, hosts=%v, groups=%v, profile=%s, signingConfig=%+v\n", commonName, hosts, groups, profile, signingConfig)
+func generateCert(
+	commonName string, hosts []string, groups []string, profile string,
+	signingConfig *config.Signing, ca []byte, caKey []byte,
+) (certBundle, error) {
+	fmt.Printf("cert generate, commonName=%s, hosts=%v, groups=%v, profile=%s, signingConfig=%+v\n",
```
Comment (ianbillett) [inline test/testtls/generate.go:132]: Preferably we would use a real logger passed to us, than a `fmt.Println` :)

File: test/e2e/services.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+	logsExtEndpoint string,
+) {
+	loki := newLokiService(e)
+	testutil.Ok(t, e2e.StartAndWaitReady(loki))
+
+	return loki.InternalEndpoint("http"),
+		loki.Endpoint("http")
+}
+
+// Starts and waits until all base services required for test are ready.
```
Comment (ianbillett) [inline test/e2e/services.go:56]: nit: comment needs to start with the function name

File: Makefile [hunk 4]
```diff
@@ -97,15 +77,19 @@ lint: $(GOLANGCI_LINT) deps shellcheck jsonnet-fmt
...
+
+.PHONY: test-e2e
+test-e2e: container
+	CGO_ENABLED=1 GO111MODULE=on go test -v -race -short -tags integration ./test/e2e
 
-.PHONY: test-integration
-test-integration: build integration-test-dependencies generate-cert
-	THANOS=$(THANOS) UP=$(UP) DEX=$(DEX) LOKI=$(LOKI) WEBSOCAT=$(WEBSOCAT) OPA=$(OPA) GUBERNATOR=$(GUBERNATOR) ./test/integration.sh
+.PHONY: test-interactive
+test-interactive: container
```
Comment (ianbillett) [inline Makefile:91]: What a great idea - so many time I just put `time.sleep(10 * time.Minute)` into my tests :smile:

File: test/config/hashrings.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
             "1610b0c3-c509-4592-a256-a1871353dbfa"
         ],
         "endpoints": [
```
Comment (ianbillett) [inline test/config/hashrings.json:7]: Can this file be delete if we are defining & running all e2e tests from Go?

File: test/e2e/helpers.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+func getContainerName(testType testType, serviceName string) string {
+	switch testType {
+	case logs:
+		return envLogsName + "-" + serviceName
+	case metrics:
+		return envMetricsName + "-" + serviceName
+	case interactive:
+		return envInteractive + "-" + serviceName
+	default:
+		return "not_found"
```
Comment (ianbillett) [inline test/e2e/helpers.go:93]: It feels more sensible to return an actual error instead of just `not_found` string?

File: test/e2e/services.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+func startServicesForLogs(t *testing.T, e e2e.Environment) (
+	logsEndpoint string,
+	logsExtEndpoint string,
+) {
+	loki := newLokiService(e)
+	testutil.Ok(t, e2e.StartAndWaitReady(loki))
+
+	return loki.InternalEndpoint("http"),
+		loki.Endpoint("http")
```
Comment (ianbillett) [inline test/e2e/services.go:53]: Is it just me or do these indents look like double tabs? They seem way bigger than normal :sweat_smile:

**Review by ianbillett — APPROVED** [review]
Run the tests locally and they all passed :tada: 

My only request would be to clean up any references to the old integration testing framework :+1:

File: test/integration.sh (hunk 1 above)
Comment (matej-g) [reply test/integration.sh:120]: Yes, I thought I removed it together with `run-local.sh` :thinking: Must be still here by mistake. Removing.

File: Makefile (hunk 4 above)
Comment (matej-g) [reply Makefile:91]: Blatantly ripped of from Thanos interactive test :stuck_out_tongue_winking_eye:

File: test/config/hashrings.json (hunk 5 above)
Comment (matej-g) [reply test/config/hashrings.json:7]: Static config files (like this one or the Loki config) from the `config` directory are still used. Only dynamic configs (e.g. requiring container hostname) are being generated during the test, static ones are copied into the shared directory.

File: test/e2e/helpers.go (hunk 6 above)
Comment (matej-g) [reply test/e2e/helpers.go:93]: Although I'm not expecting error to occur here, added a check :raised_hands:

File: test/e2e/services.go (hunk 7 above)
Comment (matej-g) [reply test/e2e/services.go:53]: Not sure :shrug: Either way that newline here is redundant.

File: test/e2e/helpers.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+	var t token
+	if err := json.Unmarshal(body, &t); err != nil {
+		return "", fmt.Errorf("cannot unmarshal token : %v\n", err)
+	}
+
+	return t.IDToken, nil
+}
+
+func getContainerName(t *testing.T, testType testType, serviceName string) string {
```
Comment (onprem) [inline test/e2e/helpers.go:84]: nit: A variable and a type with same name can be a bit confusing (`testType testType`).

File: test/e2e/helpers.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	case interactive:
+		return envInteractive + "-" + serviceName
+	default:
+		t.Fatal("invalid test type provided")
+		return ""
+	}
+}
+
+func getTLSClientConfig(t *testing.T, e e2e.Environment) *tls.Config {
+	cert, err := ioutil.ReadFile(filepath.Join(filepath.Join(e.SharedDir(), certsSharedDir, "ca.pem")))
```
Comment (onprem) [inline test/e2e/helpers.go:99]:
```suggestion
	cert, err := ioutil.ReadFile(filepath.Join(e.SharedDir(), certsSharedDir, "ca.pem"))
```

Looks like one filepath.Join is redundant.

Comment (matej-g) [reply test/e2e/helpers.go:99]: :facepalm:

**Comment by matej-g** [comment]
@ianbillett @periklis @onprem everything should be addressed now, PTAL when you get the chance, thank you :heart:

File: Makefile [hunk 10]
```diff
@@ -132,6 +116,14 @@ proto: ratelimit/gubernator/proto/google ratelimit/gubernator/gubernator.proto $
 	@cp -f $(PROTOC_GEN_GO) $(BIN_DIR)/protoc-gen-go
 	PATH=$$PATH:$(BIN_DIR):$(FIRST_GOPATH)/bin scripts/generate_proto.sh
 
+.PHONY: container-test
+container-test: build
+	@cp ./observatorium-api ./observatorium_api_for_docker
+	@docker build \
+		-f Dockerfile.e2e-test \
+		-t $(DOCKER_REPO):latest .
+	@rm ./observatorium_api_for_docker
```
Comment (periklis) [inline Makefile:125]: Why do we need this extra copy and rm here?

**Review by periklis — APPROVED** [review]
LGTM, one small nit though. Do you really format with tabs space of 2?

File: Makefile (hunk 10 above)
Comment (matej-g) [reply Makefile:125]: We don't, you're right, it's completely meaningless here, removing.

**Comment by matej-g** [comment]
> LGTM, one small nit though. Do you really format with tabs space of 2?

I'm not sure, I have my IDE set to use 4 space tabs for indentation, seems like it got changed on some files here, I fixed that. Thanks!

**Review by periklis — APPROVED** [review]
LGTM

**Pull request merged**